src/components/common/ChatBox.tsx

```diff
@@ -0,0 +1,187 @@
+/* eslint-disable arrow-body-style */
+import { RecordOf } from 'immutable';
+
+import { FC, useEffect, useRef, useState } from 'react';
+
+import { LocalContext, useLocalContext } from '@graasp/apps-query-client';
+
+import { Alert, AlertTitle, Box, Stack, styled } from '@mui/material';
+
+import { APP_DATA_TYPES, MessageData } from '../../config/appDataTypes';
+import {
+  INITIAL_PROMPT_SETTING_KEY,
+  TextResourceData,
+} from '../../config/appSettingTypes';
+import { DEFAULT_INITIAL_PROMPT } from '../../config/appSettings';
+import {
+  ANONYMOUS_USER,
+  CHATBOT_PREFIX,
+  CHATBOT_RESPONSE_URL,
+  MAX_CONVERSATION_LENGTH,
+  SCROLL_SAFETY_MARGIN,
+  STUDENT_PREFIX,
+} from '../../config/constants';
+import { CHATBOT_MODE_CY, messagesDataCy } from '../../config/selectors';
+import { LIGHT_GRAY, LIGHT_VIOLET } from '../../config/stylingConstants';
+import { useAppDataContext } from '../context/AppDataContext';
+import { useAppSettingContext } from '../context/AppSettingContext';
+import { useMembersContext } from '../context/MembersContext';
+import ChatbotBox from './ChatbotBox';
+import InputBar from './InputBar';
+import UserBox from './UserBox';
+
+const StyledMessage = styled(Box)(({ theme }) => ({
+  borderRadius: theme.spacing(1),
+  width: '70%',
+  margin: theme.spacing(1),
+  padding: theme.spacing(1),
+  flexWrap: 'wrap',
+  wordBreak: 'break-word',
+}));
+
+const StyledBotMessage = styled(StyledMessage)(() => ({
+  backgroundColor: LIGHT_VIOLET,
+}));
+
+const StyledUserMessage = styled(StyledMessage)(() => ({
+  backgroundColor: LIGHT_GRAY,
+}));
+
+type Prop = { focusWord: string; isOpen: boolean };
+
+const ChatBox: FC<Prop> = ({ focusWord, isOpen }) => {
+  const { postAppDataAsync, postAppData, appDataArray } = useAppDataContext();
+  const { appSettingArray } = useAppSettingContext();
+  const context: RecordOf<LocalContext> = useLocalContext();
+  const member = useMembersContext().find(
+    (m) => m.id === context.get('memberId'),
+  );
+  const memberName = member?.name || ANONYMOUS_USER;
+  const [loading, setLoading] = useState(false);
+
+  const initialPrompt = (
+    (appSettingArray.find((s) => s.name === INITIAL_PROMPT_SETTING_KEY)?.data ||
+      DEFAULT_INITIAL_PROMPT) as TextResourceData
+  ).text.replace('{{keyword}}', focusWord);
+
+  const chatAppData = appDataArray.filter((data) => {
+    return (
+      (data.type === APP_DATA_TYPES.BOT_COMMENT ||
+        data.type === APP_DATA_TYPES.STUDENT_COMMENT) &&
+      (data.data as MessageData).keyword === focusWord
+    );
+  });
```

Comment:
Here you should cast your result to a `(AppData & {data: MessageData})[]` which would remove the need for all the other casts that you have below.